src/ape_cache/query.py

```diff
@@ -0,0 +1,310 @@
+    AND blocks.number % :step = 0
+        """
+        ).bindparams(start_block=query.start_block, stop_block=query.stop_block, step=query.step)
+
+    @estimate_query_clause.register
+    def transaction_estimate_query_clause(self, query: BlockTransactionQuery) -> TextClause:
+        return text(
+            """
+    SELECT COUNT(*)
+    FROM transactions
+    WHERE transactions.block_hash = :block_hash
+        """
+        ).bindparams(block_hash=query.block_id)
+
+    @estimate_query_clause.register
+    def contract_events_estimate_query_clause(self, query: ContractEventQuery) -> TextClause:
+        return text(
+            """
+    SELECT COUNT(*)
+    FROM contract_events
+    WHERE contract_events.block_number >= :start_block
+    AND contract_events.block_number <= :stop_block
+    AND contract_events.block_number % :step = 0
+        """
+        ).bindparams(start_block=query.start_block, stop_block=query.stop_block, step=query.step)
+
+    @singledispatchmethod
+    def compute_estimate(self, result: CursorResult, query: QueryType) -> Optional[int]:
+        return None  # can't handle this query
+
+    @compute_estimate.register
+    def compute_estimate_block_query(
+        self,
+        query: BlockQuery,
+        result: CursorResult,
+    ) -> Optional[int]:
+        if result.rowcount == (query.stop_block - query.start_block) // query.step:
+            # NOTE: Assume 200 msec to get data from database
+            return 200
+
+        # Can't handle this query
+        # TODO: Allow partial queries
+        return None
+
+    @compute_estimate.register
+    def compute_estimate_block_transaction_query(
+        self,
+        query: BlockTransactionQuery,
+        result: CursorResult,
+    ) -> Optional[int]:
+        if result.rowcount > 0:
+            # NOTE: Assume 200 msec to get data from database
+            return 200
+
+        # Can't handle this query
+        return None
+
+    @compute_estimate.register
+    def compute_estimate_contract_events_query(
+        self,
+        query: ContractEventQuery,
+        result: CursorResult,
+    ) -> Optional[int]:
+        if result.rowcount == (query.stop_block - query.start_block) // query.step:
+            # NOTE: Assume 200 msec to get data from database
+            return 200
+
+        # Can't handle this query
+        # TODO: Allow partial queries
+        return None
+
+    def estimate_query(self, query: QueryType) -> Optional[int]:
+        try:
+            with self.database_connection as conn:
+                result = conn.execute(self.estimate_query_clause(query))
+
+                if not result:
+                    return None
+
+                return self.compute_estimate(result, query)
+        except QueryEngineError as err:
+            logger.error(f"Cannot perform query: {err}")
+            return None
+
+    # Fetch data
+    @singledispatchmethod
+    def perform_query_clause(self, query: QueryType) -> TextClause:
```

Comment:
doc string?

Reply:
currently, the core plugins are directly undocumented, besides their CLIs, so none of these will show in the docs. We would need to figure that out, probably a separate issue.

Reply:
This might be better off as an internal method anyways

Reply:
I'm resolving this conversation due to us sticking with our current documentation methodologies